### DISCUSSION TITLE
How to connect to read-write pod with RedisReplication

### DISCUSSION THREAD

**Discussion by renchap** [OT-CONTAINER-KIT/redis-operator#453] [Q&A]
Thanks a lot for this project, this is exactly what I need!

I see that you recently released the `RedisReplication`, which is perfect for us as our client does not support multiple Redis servers (yet).

The issue I am facing is that there seem to be no service targetting the current primary server, so if you connect to either of the created services, you can end up on a replica and get an error when you try to write a key.

What is the expected way for applications to have a read-write endpoint? And what is the difference between the 2 existing ones? (`<name>` and `<name>-additional`).

I would have expected the first one to point to the read-write pod, but it has 3 endpoints.

Thanks!

**Comment by shubham-cmyk** [comment]
@renchap  Right now you must setup sentinel mode along the replication to perform the write on master and read on slave

**Comment by shubham-cmyk** [comment]
I would roll out this feature in next release which is near.

**Reply by renchap** [reply]
Any news on this?

**Reply by shubham-cmyk** [reply]
@renchap You must set up 1 redis-sentinel along the replication setup or you have to pass a configuration that could allow the slaves to perform the write operation also this could be done by passing `replica-read-only no`. 

Reference docs : 
https://redis.io/docs/management/replication/

_Read-only replicas will reject all write commands, so that it is not possible to write to a replica because of a mistake._

**Reply by renchap** [reply]
I am not sure I understand your answer.

What I am looking at is simply have a Kubernetes service that points to the current write-enabled pod in a 2-pod replication setup, without Sentinel. This seems to be a very basic feature for 2-pod replication setups without Sentinel? How could you use it without this service?

**Reply by shubham-cmyk** [reply]
The problem here is the labels on the pod are identical as they are created by the statefulset. 
So pointing a fixed service to the pod becomes difficult as we have to provide some label that is only on the master redis. 

This could be achieved by using the sentinel as you forward the request to sentinel and sentinel forward to master/slave depending on the request. 

I have done some workaround but using sentinel seems to be best idea. 
By default you could point to 1 particular pod by specifying it's fqdn by service which look like <podname>.<servicename>.svc

The problem would be after failover that could be only managed by the sentinel so consider 2 case

1. No failover considered - Current service could be used as the master index is 0 and we could point to it. 
2. Failover considered - We have to use sentinel then anyway


@renchap 
If you have got any ideas i would like to hear them we could workaround that. 
Managing and label by the operator I think could be only way that you are asking

**Reply by renchap** [reply]
My issue is that some software does not support Sentinel, this is where deploying a `RedisReplication` setup can really help.

At the moment, does the operator support failing over the read-only replica? If yes, then as it makes the decision to fail over, then it should also update the service.

StatefulSets have a special label on each pod: `statefulset.kubernetes.io/pod-name=` which allows a service to point to a specific pod.
The operator could use this to create and manage a service that points to the write-enabled pod in the statefulset. If it does not handle failing over and enabling writes on the replica then this can be statically set to the first pod (write-enabled), if you handle promoting a read-replica to primary with write, it can update the service when you do the promotion.

**Reply by shubham-cmyk** [reply]
Managing the service could solve the problem we could create 1 additional service something like 
`replication-master-service`. We could use the label to detect the master pod `statefulset.kubernetes.io/pod-name=`

We would definitely introduce the failover without the sentinel mode also. So the selector on the `replication-master-service` would keep getting updated. If the failed to PING the redis-server. 

This approach could really solve the problem.

I would push this change on v0.15.1

**Reply by renchap** [reply]
Yes, you probably want 3 services:
- one targeting all pods, when you want to read on any pod
- one targeting only the primary, when you want to write
- maybe a 3rd service targeting all read-only pods, when you want to have heavy read operations and dont want to impact the primary performance

**Reply by shubham-cmyk** [reply]
Sounds Good! 3 services could be created here.

**Reply by sdwilsh** [reply]
@shubham-cmyk, is this work tracked in an issue somewhere?

**Comment by pierreozoux** [comment]
I believe, from what I understand, sentinel cluster has 2 jobs:
- promote a replica to primary
- tell clients which one is the primary or the replica

For the first question, there is an issue about that topic here:
https://github.com/OT-CONTAINER-KIT/redis-operator/issues/775

For the second question, which is exactly the question of OP, I created this issue:
https://github.com/OT-CONTAINER-KIT/redis-operator/issues/1341
